### PR TITLE
Add deprecation warning for intrinsic `__CPROVER_allocated_memory` 

### LINKED ISSUE #6872: Deprecate `__CPROVER_allocated_memory`
The plan for `__CPROVER_allocated_memory` seems to be to remove it in PR #6748 and replace it with a command line option in #6747 for the next major release of cbmc. Therefore we should emit a suitable warning about it being deprecated when it is found in user code. That way we can discourage new uses of it and any existing users will be suitably warned that it will need to be replaced.

### PR DESCRIPTION
**Pull request by NlightNFotis**
In line with https://github.com/diffblue/cbmc/issues/6872, this PR adds a deprecation warning for the usage of the intrinsic `__CPROVER_allocated_memory`.

Fixes: https://github.com/diffblue/cbmc/issues/6872

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]
You might just use one of the existing tests that currently use `__CPROVER_allocated_memory` to add a pattern.

**Comment by NlightNFotis** [comment]
@tautschnig Yeah, that's what I was planning to do :)

**Review by TGWDB — APPROVED** [review]
Happy to approve as is, but note that if we're starting to plan Version 6 changes and deprecations, we may want a macro or standard way to implement deprecation warnings.

**Comment by NlightNFotis** [comment]
@kroening Hi Daniel, this is blocked on a code owner's review from you. Could you please have a look when you find some time? Thank you 👍🏻

**Pull request merged**
